### PR TITLE
Remove unnecessary trait bounds and derivations

### PR DESCRIPTION
**Pull request by nnethercote**
This PR removes unnecessary trait bounds and derivations from many types.

r? @nikomatsakis

### CODE REVIEW

**Comment by nnethercote** [comment]
I did this because I was trying to work out why `Symbol` implemented `Hash`, and in doing so I found a bunch of types that derived `Hash` unnecessarily. Then I took a semi-mechanical approach:

- I removed `Hash` trait bounds from all types, except those in tests and `lib{alloc,core,std}`.
- I compiled (and recompiled...), and added back the `Hash` trait bounds everywhere the compiler said it was necessary.
- I then removed all `Hash` derives, except those in tests and `lib{alloc,core,std}`.
- I compiled/recompiled, adding back all `Hash` derives everywhere the compiler said was necessary.
- I figure the types with unnecessary `Hash` derives were good candidates for possibly having other unnecessary dervies. So I then removed all derives from these types, and the re-added them back in as necessary.

Overall, I did a thorough job eliminating unnecessary `Hash` bounds and derives, and a scattershot job for other trait bounds and derives.

The whole process was very tedious, but it did find a few unnecessary trait bounds and many unnecessary derives. It would be nice if the compiler could identify at least some of these, the way it identifies unnecessary `use` items.

File: src/librustc_data_structures/transitive_relation.rs [hunk 1]
```diff
@@ -11,7 +11,7 @@ use std::mem;
 mod tests;
 
 #[derive(Clone, Debug)]
-pub struct TransitiveRelation<T: Clone + Debug + Eq + Hash> {
+pub struct TransitiveRelation<T: /*Clone + Debug +*/ Eq + Hash> {
```
Comment (Centril) [inline src/librustc_data_structures/transitive_relation.rs]: Why comment?

Comment (nnethercote) [reply src/librustc_data_structures/transitive_relation.rs]: Whoops, good catch. I have fixed and updated.

**Comment by Centril** [comment]
r? @Centril @bors r+

-- Looks great; some tools may break as a result of this but we can re-add specific bounds as needed.

**Comment by matthiaskrgr** [comment]
Clippy needs some of these bounds.

````
error[E0369]: binary operation `==` cannot be applied to type `syntax::ast::LitKind`
   --> clippy_lints/src/utils/hir_utils.rs:117:70
    |
117 |             (&ExprKind::Lit(ref l), &ExprKind::Lit(ref r)) => l.node == r.node,
    |                                                               ------ ^^ ------ syntax::ast::LitKind
    |                                                               |
    |                                                               syntax::ast::LitKind
    |
    = note: an implementation of `std::cmp::PartialEq` might be missing for `syntax::ast::LitKind`

error[E0599]: no method named `hash` found for type `&syntax::source_map::Spanned<rustc::hir::BinOpKind>` in the current scope
   --> clippy_lints/src/utils/hir_utils.rs:414:19
    |
414 |                 o.hash(&mut self.s);
    |                   ^^^^ method not found in `&syntax::source_map::Spanned<rustc::hir::BinOpKind>`
    |
    = note: the method `hash` exists but the following trait bounds were not satisfied:
            `&syntax::source_map::Spanned<rustc::hir::BinOpKind> : std::hash::Hash`

error[E0599]: no method named `hash` found for type `rustc::hir::BinOpKind` in the current scope
   --> clippy_lints/src/utils/hir_utils.rs:422:25
    |
422 |                 op.node.hash(&mut self.s);
    |                         ^^^^ method not found in `rustc::hir::BinOpKind`

error[E0599]: no method named `hash` found for type `&syntax::source_map::Spanned<syntax::ast::LitKind>` in the current scope
   --> clippy_lints/src/utils/hir_utils.rs:463:19
    |
463 |                 l.hash(&mut self.s);
    |                   ^^^^ method not found in `&syntax::source_map::Spanned<syntax::ast::LitKind>`
    |
    = note: the method `hash` exists but the following trait bounds were not satisfied:
            `&syntax::source_map::Spanned<syntax::ast::LitKind> : std::hash::Hash`

error[E0599]: no method named `hash` found for type `rustc::hir::UnOp` in the current scope
   --> clippy_lints/src/utils/hir_utils.rs:522:21
    |
522 |                 lop.hash(&mut self.s);
    |                     ^^^^ method not found in `rustc::hir::UnOp`

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0369, E0599.
For more information about an error, try `rustc --explain E0369`.
error: could not compile `clippy_lints`.
warning: build failed, waiting for other jobs to finish...
error[E0369]: binary operation `==` cannot be applied to type `syntax::ast::LitKind`
   --> clippy_lints/src/utils/hir_utils.rs:117:70
    |
117 |             (&ExprKind::Lit(ref l), &ExprKind::Lit(ref r)) => l.node == r.node,
    |                                                               ------ ^^ ------ syntax::ast::LitKind
    |                                                               |
    |                                                               syntax::ast::LitKind
    |
    = note: an implementation of `std::cmp::PartialEq` might be missing for `syntax::ast::LitKind`

error[E0599]: no method named `hash` found for type `&syntax::source_map::Spanned<rustc::hir::BinOpKind>` in the current scope
   --> clippy_lints/src/utils/hir_utils.rs:414:19
    |
414 |                 o.hash(&mut self.s);
    |                   ^^^^ method not found in `&syntax::source_map::Spanned<rustc::hir::BinOpKind>`
    |
    = note: the method `hash` exists but the following trait bounds were not satisfied:
            `&syntax::source_map::Spanned<rustc::hir::BinOpKind> : std::hash::Hash`

error[E0599]: no method named `hash` found for type `rustc::hir::BinOpKind` in the current scope
   --> clippy_lints/src/utils/hir_utils.rs:422:25
    |
422 |                 op.node.hash(&mut self.s);
    |                         ^^^^ method not found in `rustc::hir::BinOpKind`

error[E0599]: no method named `hash` found for type `&syntax::source_map::Spanned<syntax::ast::LitKind>` in the current scope
   --> clippy_lints/src/utils/hir_utils.rs:463:19
    |
463 |                 l.hash(&mut self.s);
    |                   ^^^^ method not found in `&syntax::source_map::Spanned<syntax::ast::LitKind>`
    |
    = note: the method `hash` exists but the following trait bounds were not satisfied:
            `&syntax::source_map::Spanned<syntax::ast::LitKind> : std::hash::Hash`

error[E0599]: no method named `hash` found for type `rustc::hir::UnOp` in the current scope
   --> clippy_lints/src/utils/hir_utils.rs:522:21
    |
522 |                 lop.hash(&mut self.s);
    |                     ^^^^ method not found in `rustc::hir::UnOp`

error: aborting due to 5 previous errors

````

Can we revert this?

**Comment by lzutao** [comment]
Just add necessary bound for clippy I think.

**Comment by matthiaskrgr** [comment]
Can we do this for external types?

**Comment by lzutao** [comment]
I mean re-add those bound in rust repo, not in clippy side.

**Comment by mati865** [comment]
Maybe Clippy could use Hashstable instead of Hash.
Probably PartialEq is unavoidable.

**Comment by eddyb** [comment]
@nnethercote Looks like this caused a lot of fallout - why not just remove `Hash` from `Symbol` and see what breaks?

**Pull request merged**